helsingborg-stad/helsingborg-io-sls-api
# Fix/completions_redesign_flow

## Explain the changes you’ve made

Created lambda - `syncCaseCompletions` whose task is to retrieve data for completion from Vada and update the attribute` completions` on the case. The lambda is triggered at the event: `applicationStatusSuccess`

Updated the lambda - `setCaseCompletions` whose task is to update the attributes `status.type`, `state` and` currentFormId`. The lambda sets `person.hasSigned` to `false` to be able to sign the completion form again.

Updated the lambda - `checkCompletionsStatus` whose task is to check if Viva status code is 64 - completions. The Lambda is triggered on the event: `syncSuccess`. If Viva status code is not 64 then we have to check if the completion due date is expired because even if the Viva status code lacks the status code 64, the workflow could still be in a completions state.

Created lambda - `checkCompletionsDueDate` whose task is to check if the due date for completion has passed.
The lambda is triggered on event: `checkCompletionSuccess` and only if `state` on the case is set to `VIVA_RANDOM_CHECK_REQUIRED` or `VIVA_COMPLETION_REQUIRED`.

## Explain why these changes are made

These changes complete the completions experience

## How to test

1. Checkout this branch
2. Make sure your AWS sandbox is **up to date with all services** from this branche
3. run `sls deploy --force` within service folder for `service viva-ms`
4. Have one open period ready in Viva-admin for an Viva test user

Expected status code from Vada endpoint:

`https://dev-viva-adapter.helsingborg.se/applications/[HASHID FOR YOUR VIVA TEST USER]/status`

```json
[ { "code": 128, "description": "Ärende finns (försörjningsstödsärende)" }, { "code": 256, "description": "Ett ärende är aktiverat på webben" }, { "code": 512, "description": "Ärendet tillåter e-ansökan" }, { "code": 1, "description": "Ansökan tillåten" } ]
```

5. Complete the application using the app or Postman or equivalent if you feel adventurous

**Submit case using Postman**

**PUT**

url: https://_[URL TO YOUR AWS SANDBOX]_/cases/_[CASE ID OF THE OPEN APPICATION PERIOD I DYNAMODB CASES TABLE]_

Headers:

- Authorization: [TEST VIVA USER TOKEN]
- x-api-key: [VADA DEV TOKEN]

Payload
```json
{ "currentFormId": "a1ce65a0-3d7c-11ec-801e-2db8270ee167", "currentPosition": { "currentMainStepIndex": 12, "index": 12, "level": 0, "currentMainStep": 7 }, "answers": [ { "field": { "id": "rent", "tags": [ "expenses", "boende", "amount" ] }, "value": "12580,50" } ], "encryption": { "type": "decrypted" }, "signature": { "success": true } }
```

There is a 50-50 chance that the application will end up in a random check. If not, you can submit a new application again with the new case ID created in the database

6.  Alter the completions due date in Viva-admin by entering an date before todays date
7.  Login via the app and confirm that the case is in the state _Behandlas_ and that the dud date is not pressent

## Tested environments


---

## Checklist for reviewers

Here are a few points that are useful for reviewers to check.

The short list of the (maybe) most important points are:

- Does this PR work as expected?
- Does the PR properly only concern a single feature/bug/thing?
- Is the code clear and concise?
- Will the changes impact existing functionality?
- Are there any linting errors (ESLint) that should be resolved?
- Do the automatic GitHub checks pass?

However it is good if a reviewer can be as thorough as possible,
within reasonable limits of course.

### Functionality

- Does this PR work as expected?
- If this is a new feature, does it add value, or is it a sign of feature-creep?
- Will the changes impact existing functionality?
- Does the changes create inconsistencies somewhere else?
- Does the code work on different platforms if applicable?
- How does the code handle if unexpected values are passed to functions etc.?

### Readability, syntax, formatting, style

- Is the code clear and concise?
- Has styling rules (Prettier) been applied to included files?
- Are functions/variables/etc. aptly named?

### Design

- Is the code properly planned and designed?
- Is code organized/separated in a logical fashion?
- Does the code follow clean code practices to a reasonable degree?
- Is the code self-documenting enough?
- Is the code in sync with existing patterns/technologies?
- Does the code reuse existing functionality where applicable?
- Is the code designed with reusability in mind?
- Does the code have hard-coded values that can be refactored?

### Patterns, idioms, tests, best practices

- Are there any linting errors (ESLint) that should be resolved?
- Is the code properly typed (TypeScript)?
- Does the code utilize language features to their full potential?

### Debugging, testing, configuring

- Is appropriate logging used?
- Have tests been added/modified to accommodate the changes?
- Is the feature sufficiently configurable in case tweaks are required in the future?

### Performance, reliability, scalability

- Is the code well optimized?
- Does the code scale?
- Does the code handle failure scenarios?

### Security

- Is there any obvious potential for exploits and security threats?
- Does code check proper authorization where applicable?

### Etiquettes

- Are commits atomic (does only one thing per commit)?
- Does the PR properly only concern a single feature/bug/thing?
- Are commit messages well-written?
- Do the automatic GitHub checks pass?
